components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java

```diff
@@ -4393,4 +4402,84 @@ public static String resolveTenantDomain(HttpServletRequest request) {
+            throws IdentityOAuth2Exception {
+        org.wso2.carbon.user.api.UserStoreManager userStoreManager;
+        UserRealm userRealm;
+        try {
+            int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
+            RealmService realmService = IdentityTenantUtil.getRealmService();
+            userRealm = realmService.getTenantUserRealm(tenantId);
+            userStoreManager = userRealm.getUserStoreManager();
+        } catch (UserStoreException e) {
+            throw new IdentityOAuth2Exception("Error occurred while loading user manager from user realm", e);
+        }
+
+        String passwordLastChangedTime;
+        String claimURI = PasswordPolicyConstants.LAST_CREDENTIAL_UPDATE_TIMESTAMP_CLAIM;
+        try {
+            passwordLastChangedTime = getLastPasswordUpdateTime(userStoreManager, claimURI, tenantAwareUsername);
+            if (passwordLastChangedTime == null) {
+                ClaimManager claimManager = userRealm.getClaimManager();
+                claimURI = PasswordPolicyConstants.LAST_CREDENTIAL_UPDATE_TIMESTAMP_CLAIM_NON_IDENTITY;
+                if (claimManager.getClaim(claimURI) != null) {
+                    passwordLastChangedTime =
+                            getLastPasswordUpdateTime(userStoreManager, claimURI, tenantAwareUsername);
+                }
+            }
+        } catch (UserStoreException e) {
+            throw new IdentityOAuth2Exception("Error occurred while loading user claim - " + claimURI, e);
+        }
+
+        long passwordChangedTime = 0;
+        if (passwordLastChangedTime != null) {
+            passwordChangedTime = Long.parseLong(passwordLastChangedTime);
+        }
+
+        int daysDifference = 0;
+        long currentTimeMillis = System.currentTimeMillis();
+        if (passwordChangedTime > 0) {
+            Calendar currentTime = Calendar.getInstance();
+            currentTime.add(Calendar.DATE, (int) currentTime.getTimeInMillis());
+            daysDifference = (int) ((currentTimeMillis - passwordChangedTime) / (1000 * 60 * 60 * 24));
+        }
+
+        int passwordExpiryInDays = PasswordPolicyConstants.PASSWORD_EXPIRY_IN_DAYS_DEFAULT_VALUE;
+
+        // Getting the configured number of days before password expiry in days
+        String passwordExpiryInDaysConfiguredValue = PasswordPolicyUtils
+                .getResidentIdpProperty(tenantDomain, PasswordPolicyConstants.PASSWORD_EXPIRY_IN_DAYS_FROM_CONFIG);
+
+        if (StringUtils.isEmpty(passwordExpiryInDaysConfiguredValue)) {
+            passwordExpiryInDaysConfiguredValue = PasswordPolicyUtils.getIdentityEventProperty(
+                    PasswordPolicyConstants.PASSWORD_EXPIRY_IN_DAYS_FROM_CONFIG);
+        }
+
+        if (passwordExpiryInDaysConfiguredValue != null) {
+            passwordExpiryInDays = Integer.parseInt(passwordExpiryInDaysConfiguredValue);
+        }
+
+        return (daysDifference > passwordExpiryInDays || passwordLastChangedTime == null);
```

Comment:
For a case where this feature is gonna be enabled after applying the update; As we are subscribing to ADD password event with this fix, we can assume that currently, existing users may not have a value for this claim? Isn't it? If so the users who do not have the last password update claim has to be considered as valid users who do not have had their password expired.